File: apps/web/src/server/optimizer/requirements.ts
import type { RequestRequirements, TaskType } from "./types";

type ChatMessage = {
  role?: string;
  content?: unknown;
  tool_calls?: unknown;
};

type ChatBody = {
  messages?: ChatMessage[];
  tools?: unknown[];
  functions?: unknown[];
  response_format?: { type?: string } | null;
  max_tokens?: number;
  max_completion_tokens?: number;
};

function contentHasImage(content: unknown): boolean {
  if (!Array.isArray(content)) return false;
  return content.some((part) => {
    if (!part || typeof part !== "object") return false;
    const p = part as { type?: string; image_url?: unknown };
    return p.type === "image_url" || Boolean(p.image_url);
  });
}

function estimateTokens(text: string) {
  return Math.max(1, Math.round(text.length / 4));
}

function flattenContent(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .map((part) => {
        if (typeof part === "string") return part;
        if (part && typeof part === "object" && "text" in part) return String((part as { text: unknown }).text ?? "");
        return "";
      })
      .join("\n");
  }
  return "";
}

function inferTaskType(blob: string): TaskType {
  if (/```|def |function |class |bug|stack trace|typescript|python/i.test(blob)) return "coding";
  if (/prove|theorem|reason step|chain of thought|logic puzzle/i.test(blob)) return "reasoning";
  if (/extract|json schema|fill the fields|structured/i.test(blob)) return "extraction";
  if (/summarize|tl;dr|synopsis/i.test(blob)) return "summarization";
  if (/what is|who is|capital of|when did/i.test(blob)) return "factual_qa";
  if (blob.length > 12_000) return "long_context";
  return "general";
}

export function extractRequirements(body: ChatBody): RequestRequirements {
  const messages = body.messages ?? [];
  const blob = messages.map((m) => flattenContent(m.content)).join("\n");
  const hasTools = Boolean(body.tools?.length || body.functions?.length || messages.some((m) => m.tool_calls));
  const hasVision = messages.some((m) => contentHasImage(m.content));
  const structured =
    body.response_format?.type === "json_object" ||
    body.response_format?.type === "json_schema" ||
    /return (only )?json|respond with json|output a json/i.test(blob);
  const taskType = inferTaskType(blob);
  const contextTokens = estimateTokens(blob) + 512;
  const minOut = body.max_completion_tokens ?? body.max_tokens ?? 0;

  return {
    requires_tools: hasTools,
    requires_reasoning: taskType === "reasoning" || /step by step|show your work/i.test(blob),
    requires_structured_output: structured,
    requires_vision: hasVision || taskType === "vision",
    minimum_context_tokens: contextTokens,
    minimum_output_tokens: minOut,
    task_type: hasVision ? "vision" : taskType,
  };
}

File: apps/api/app/domain/optimizer/requirements.py
from __future__ import annotations

from typing import Any

from app.domain.optimizer.schemas import RequestRequirements, TaskType


def extract_requirements(
    messages: list[dict[str, Any]],
    extra: dict[str, Any] | None = None,
) -> RequestRequirements:
    extra = extra or {}
    parts: list[str] = []
    has_vision = False
    for message in messages:
        content = message.get("content")
        if isinstance(content, str):
            parts.append(content)
        elif isinstance(content, list):
            for block in content:
                if isinstance(block, dict):
                    if block.get("type") == "image_url" or block.get("image_url"):
                        has_vision = True
                    if block.get("text"):
                        parts.append(str(block["text"]))
    blob = "\n".join(parts)
    rf = extra.get("response_format") or {}
    structured = isinstance(rf, dict) and rf.get("type") in {
        "json_object",
        "json_schema",
    }
    tools = bool(extra.get("tools") or extra.get("functions"))
    task = TaskType.FACTUAL_QA
    lower = blob.lower()
    if "```" in blob or "def " in blob or "function " in lower:
        task = TaskType.CODING
    elif "prove" in lower or "reason step" in lower:
        task = TaskType.REASONING
    elif has_vision:
        task = TaskType.VISION
    elif "summarize" in lower:
        task = TaskType.SUMMARIZATION
    max_out = extra.get("max_tokens") or extra.get("max_completion_tokens") or 0
    return RequestRequirements(
        requires_tools=tools,
        requires_reasoning=task == TaskType.REASONING,
        requires_structured_output=structured or "respond with json" in lower,
        requires_vision=has_vision,
        minimum_context_tokens=max(0, len(blob) // 4 + 512),
        minimum_output_tokens=int(max_out),
        task_type=TaskType.VISION if has_vision else task,
    )
